processor/lsmintervalprocessor/processor.go

```diff
@@ -0,0 +1,388 @@
+		processingTime: time.Now().UTC().Truncate(cfg.Intervals[0]),
+		ctx:            ctx,
+		cancel:         cancel,
+		logger:         log,
+	}, nil
+}
+
+func (p *Processor) Start(ctx context.Context, host component.Host) error {
+	p.mu.Lock()
+	if p.db == nil {
+		db, err := pebble.Open(p.dataDir, p.dbOpts)
+		if err != nil {
+			return fmt.Errorf("failed to open database: %w", err)
+		}
+		p.db = db
+	}
+
+	if p.exportStopped == nil {
+		p.exportStopped = make(chan struct{})
+	}
+	p.mu.Unlock()
+
+	go func() {
+		defer close(p.exportStopped)
+		to := p.processingTime.Add(p.intervals[0])
+		timer := time.NewTimer(time.Until(to))
+		defer timer.Stop()
+
+		for {
+			select {
+			case <-p.ctx.Done():
+				return
+			case <-timer.C:
+			}
+
+			p.mu.Lock()
+			batch := p.batch
+			p.batch = nil
+			p.processingTime = to
+			p.mu.Unlock()
+
+			// Export the batch
+			if err := p.commitAndExport(p.ctx, batch, to); err != nil {
+				p.logger.Warn("failed to export", zap.Error(err), zap.Time("end_time", to))
+			}
+
+			to = to.Add(p.intervals[0])
+			timer.Reset(time.Until(to))
+		}
+	}()
+	return nil
+}
+
+func (p *Processor) Shutdown(ctx context.Context) error {
+	// Signal stop for the exporting goroutine
+	p.cancel()
+
+	// Wait for the exporting goroutine to stop. Note that we don't need to acquire
+	// mutex here since even if there is a race between Start and Shutdown the
+	// processor context is cancelled ensuring export goroutine will be noop.
+	if p.exportStopped != nil {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("failed to shutdown due to context timeout while waiting for export to stop: %w", ctx.Err())
+		case <-p.exportStopped:
+		}
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// Ensure all data in the database is exported
+	if p.db != nil {
+		p.logger.Info("exporting all data before shutting down")
+		if p.batch != nil {
+			if err := p.batch.Commit(p.wOpts); err != nil {
+				return fmt.Errorf("failed to commit batch: %w", err)
+			}
+			if err := p.batch.Close(); err != nil {
+				return fmt.Errorf("failed to close batch: %w", err)
+			}
+			p.batch = nil
+		}
+
+		var errs []error
+		for _, ivl := range p.intervals {
+			// At any particular time there will be 1 export candidate for
+			// each aggregation interval. We will align the end time and
+			// process each of these.
+			to := p.processingTime.Truncate(ivl).Add(ivl)
+			if err := p.export(ctx, to); err != nil {
+				errs = append(errs, fmt.Errorf(
+					"failed to export metrics for interval %s: %w", ivl, err),
+				)
+			}
+		}
+		if len(errs) > 0 {
+			return fmt.Errorf("failed while running final export: %w", errors.Join(errs...))
+		}
+		if err := p.db.Close(); err != nil {
+			return fmt.Errorf("failed to close database: %w", err)
+		}
+		// All future operations are invalid after db is closed
+		p.db = nil
+	}
+	return nil
+}
+
+func (p *Processor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: true}
```

Comment:
I wonder if the "MutatesData" capability is a bit too coarse-grained. IIANM we're only removing metrics from the slice, but we don't otherwise modify the metric contents, and this capability would cause them all to be cloned I think.

Maybe something to keep an eye on in benchmarks, and open an issue about upstream.